airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -471,6 +463,102 @@ String insertNewRecords(final StreamConfig stream,
+                                  final String finalSuffix,
+                                  final boolean forceSafeCasting,
+                                  final Optional<Instant> minRawTimestamp) {
+    final String pkEquivalent = stream.primaryKey().stream().map(pk -> {
+      final String quotedPk = pk.name(QUOTE);
+      // either the PKs are equal, or they're both NULL
+      return "(target_table." + quotedPk + " = new_record." + quotedPk
+          + " OR (target_table." + quotedPk + " IS NULL AND new_record." + quotedPk + " IS NULL))";
+    }).collect(joining(" AND "));
+
+    final String columnList = stream.columns().keySet().stream()
+        .map(quotedColumnId -> quotedColumnId.name(QUOTE) + ",")
+        .collect(joining("\n"));
+    final String newRecordColumnList = stream.columns().keySet().stream()
+        .map(quotedColumnId -> "new_record." + quotedColumnId.name(QUOTE) + ",")
+        .collect(joining("\n"));
+    final String extractNewRawRecords = extractNewRawRecords(stream, forceSafeCasting, minRawTimestamp);
+
+    final String cursorComparison;
+    if (stream.cursor().isPresent()) {
+      final String cursor = stream.cursor().get().name(QUOTE);
+      // Build a condition for "new_record is more recent than target_table":
+      cursorComparison =
+          // First, compare the cursors.
+          "(target_table." + cursor + " < new_record." + cursor
+          // Then, break ties with extracted_at. (also explicitly check for both new_record and final table
+          // having null cursor
+          // because NULL != NULL in SQL)
+              + " OR (target_table." + cursor + " = new_record." + cursor
+              + " AND target_table._airbyte_extracted_at < new_record._airbyte_extracted_at)"
+              + " OR (target_table." + cursor + " IS NULL AND new_record." + cursor
+              + " IS NULL AND target_table._airbyte_extracted_at < new_record._airbyte_extracted_at)"
+              // Or, if the final table has null cursor but new_record has non-null cursor, then take the new
+              // record.
+              + " OR (target_table." + cursor + " IS NULL AND new_record." + cursor + " IS NOT NULL))";
+    } else {
+      // If there's no cursor, then we just take the most-recently-emitted record
+      cursorComparison = "target_table._airbyte_extracted_at < new_record._airbyte_extracted_at";
+    }
+
+    final String cdcDeleteClause;
+    final String cdcSkipInsertClause;
+    if (stream.columns().containsKey(CDC_DELETED_AT_COLUMN)) {
+      // Execute CDC deletions if there's already a record
+      cdcDeleteClause = "WHEN MATCHED AND new_record._ab_cdc_deleted_at IS NOT NULL AND " + cursorComparison + " THEN DELETE";
+      // And skip insertion entirely if there's no matching record.
+      // (This is possible if a single T+D batch contains both an insertion and deletion for the same PK)
+      cdcSkipInsertClause = "AND new_record._ab_cdc_deleted_at IS NULL";
```

Comment:
To confirm my understanding, in the scenario of `WHEN NOT MATCHED AND new_record._ab_cdc_deleted_at IS NOT NULL`, we don't insert (or do anything else) bc the record has been deleted?

I'm also confused by the naming `cdcSkipInsertClause` which is then immediately followed by an `INSERT` on L549.

Reply:
> confirm my understanding

correct. E.g. if a single T+D batch includes a record for `INSERT INTO users ...` and also `DELETE FROM users ...`, then `new_record` will have deleted_at=<something>, and we don't want to insert it.

Reply:
> cdcSkipInsertClause

yeah I'm not great at naming :/ how about... `cdcDeletionClause`? idk